### ISSUE TITLE
refactor: Move Code to Typescript

### ISSUE THREAD

**Issue by wuda-io** [materializecss/materialize#267]
For cleaner development and also to go with the time it is greatly apprecaiated to move the Codebase to TypeScript.

For consumers of Materialize everything stays the same, so no worries.
The big advantage is for developing the Framework is to have a much better codebase and also reduce the risk of errors,
since the typing system of TypeScript enhances the development experience.

The steps would be:
1. Convert existing JS-Files to TS Files (since TS is a sub-set of JS)
2. Setup the CI-Process to compile the TS-Files down to a single JS-File or to Components (please comment, help appreciated here)
3. Refactor existing Code and Setup Classes, Types and interfaces as needed

I think we should compile using the ES06 Version of JavaScript.
What do you think?

**Comment by DanielRuf** [comment]
@materializecss/members

**Comment by LoganTann** [comment]
This need a refactor of the whole codebase. Of course, we will use it once someone is ready to start v2.

But for v1 (which only fixes bug issues), the only alternative is using JsDoc types.

**Comment by coffeeandwork** [comment]
> This need a refactor of the whole codebase. Of course, we will use it once someone is ready to start v2.
> 
> But for v1 (which only fixes bug issues), the only alternative is using JsDoc types.

Is there a doc somewhere that talks about what might go into version 2? I looked through the discussions and I could not find one. I did find a couple of closed projects:

- [https://github.com/materializecss/materialize/projects/3](https://github.com/materializecss/materialize/projects/3)
- [https://github.com/materializecss/materialize/projects/2](https://github.com/materializecss/materialize/projects/2)

I am guessing the projects were closed because there was no activity on them? Would it make sense to create a project (or doc) and start to add tasks to it to get a full picture of what version 2 could look like (if there isn't one already that I haven't found)? I would like to help out if I can.

**Comment by ChildishGiant** [comment]
There's a milestone for 2.0 https://github.com/materializecss/materialize/milestone/1

**Comment by danice** [comment]
Hello!
I have done a proof of moving some components to typescript. 
See [here](https://github.com/danice/materializecss-typescript-proof) is the project.

This project includes:
- moving Tabs and Carousel components to typescript
- bundling them with webpack to materialize-css.js file
- loading the bundle from a regular html file and initializing the components

I would say that very few changes where required to move the components (you can see changes needed for tab component [here](https://github.com/danice/materializecss-typescript-proof/commit/17fe2d527858f5210d44e5ef7149d14df8aef033#diff-407d55845937db1cd9a25180516dadb4febe20d3b28dc48cf78bfa12c0e9663e]).

I have not bundled the sass files, but I don't think it would give problems.

I think with we can use this as starting point to move the code to typescript and it would not be very difficult.
What do you think? Any ideas or suggestions?

**Comment by roiLeo** [comment]
Hey, nice work!

> What do you think? Any ideas or suggestions?

I would suggest some things:
- avoid using too often `any` Type
- use `const` instead of `let`
- remove jQuery!

**Comment by wuda-io** [comment]
Hey @danice 
That looks awesome. You definitely proofed that it would work!
I think it is a lot to refactor, but I would like to review the changes if we can progress within a PR
step by step and also keeping all the functionality.
We could first start with the CI and then progress to using clean types and interfaces.

Also I agree with @roiLeo and would also remove jQuery if there is not too much damage during the refacoring.
We should consider https://youmightnotneedjquery.com/

**Comment by danice** [comment]
The jquery was just a confusion, it works just the same with cash-dom. I have just updated the code without jquery dependency.

The "any" fields are because I used the editor to create the missing properties. I have not even used the param comments to define function parameter types, so differences with js code be minimum.

**Comment by danice** [comment]
I see that the framework uses Grunt to process the files. I have no experience processing typescript with Grunt, but I can try it, to minimize CI changes.

**Comment by danice** [comment]
List of advances in proof project:
- grunt task executing webpack 
- autocomplete and dropdown converted (dropdown not tested)
- component intialization with  M.AutoInit();
- component initiallization with jquery
- M compo renamed to Global to avoid confusion with library name

**Comment by danice** [comment]
Hi!
I has been more work than expected but a last the first version of [typescript migration is finished](https://github.com/danice/materialize/tree/move-to-ts) 

The objectives of this first step was:
- just convert all files to typescript maintaining everything as is and with minimal changes in the code so it can be quickly review
This means I have not dedicated to try add all type definitions, just the ones automatically added and some required ones. 

- update the grunt process to use webpack instead of babel, but again maintaining everything as simple as possible.

Pending tasks:
1)Tests: There are about 3-4 test that I cannot get to work. I think I have been able to fix the tested functionality, but not in the test.
As I have not been able to debug the tests is difficult for me to found the reason. 

2) complete types for components. Add types for every component option. 

3) someone who knows better the framework should check that grunt processes work completely as expected and a complete checking that  components can we loaded and works just as before.
